File: backend/src/routes/positions.py
from fastapi import APIRouter, Depends
from src.services.users import UserService
from src.containers import Container
from src.services.positions import PositionService
from src.models.position import PositionCreate
from dependency_injector.wiring import inject, Provide

router = APIRouter(
    prefix="/positions", dependencies=[Depends(Provide[Container.position_service])]
)


@router.get("/{position_id}")
@inject
async def getPosition(
    position_id,
    position_service: PositionService = Depends(Provide[Container.position_service]),
):
    position = position_service.get(position_id)
    return {
        "name": position.position_type.name,
        "user": position.user,
        "id": position.id,
        "paid": position.paid,
    }


# TODO: Refactor Position and Position endpoints
@router.post("", status_code=201)
@inject
async def createPosition(
    position: PositionCreate,
    position_service: PositionService = Depends(Provide[Container.position_service]),
):
    position_service.create(position)


@router.post("/{position_id}/user/{user_id}")
@inject
async def takePosition(
    position_id,
    user_id,
    position_service: PositionService = Depends(Provide[Container.position_service]),
    user_service: UserService = Depends(Provide[Container.user_service]),
):
    position_service.take_position(position_id, user_id)
    position = position_service.get(position_id)
    return {
        "name": position.position_type.name,
        "user": position.user,
        "id": position.id,
    }


@router.delete("/{position_id}/user")
@inject
async def releasePosition(
    position_id,
    position_service: PositionService = Depends(Provide[Container.position_service]),
):
    position_service.release_position(position_id)

File: frontend/src/lib/requests/positions.ts
import client from "../client"
import { Position } from "../models/Position"
import { PositionType } from "../models/PositionType"


export async function loadPositionTypes() {
  const endpoint = `/position-types`

  return (await client.get<PositionType[]>(endpoint)).data
}

export async function getPosition(positionId: string) {
  const endpoint = `/positions/${positionId}`

  return (await client.get<Position>(endpoint)).data
}

export async function getUserPositions(userId: string) {
  const endpoint = `/positions?userId=${userId}`

  return (await client.get<Position>(endpoint)).data
}

export async function createPosition(name: string) {
  const endpoint = `/positions`

  await client.post(endpoint, { name })
}

export async function createPositionType(name: string) {
  const endpoint = `/position-types`

  await client.post(endpoint, { name })
}

export async function deletePosition(positionId: string) {
  const endpoint = `/positions/${positionId}`

  await client.delete(endpoint)
}

export async function addPosition(teamId: string, positionTypeId: string) {
  const endpoint = `/teams/${teamId}/positions`

  await client.post(endpoint, { positionTypeId })
}

export async function takePositionWithUser(positionId: string, userId: string) {
  const endpoint = `positions/${positionId}/user/${userId}`

  return (await client.post(endpoint)).data
}

export async function releasePosition(positionId: string) {
  const endpoint = `positions/${positionId}/user`

  return (await client.delete(endpoint)).data
}

export async function takePositionWithName(eventId: string, teamId: string, positionId: string, userName: string) {
  const endpoint = `events/${eventId}/teams/${teamId}`

  client.put(endpoint, {
    positionId: positionId,
    userName: userName
  })
}

export async function markPositionPaid(positionId: string) {
  const endpoint = `/positions/${positionId}/paid`

  return await client.put(endpoint)
}